csrc/multidevice/communicator.cpp

```diff
@@ -214,7 +214,9 @@ Communicator::Communicator(
 #endif
 }
 
-Communicator::~Communicator() {
+void Communicator::cleanup() {
```

Comment:
Why  does it make a difference to name the method `cleanup` instead of being the destructor? As far as I can see we could use the destructor in place of cleanup, without any change of behavior, we would just use `delete Communicator::getInstance();` instead of `Communicator::getInstance().cleanup();`

If I'm missing something and we can't use the destructor, then it's maybe worth adding in the destructor a check that cleanup has been called before, and print a warning otherwise.

Reply:
It's harder to detect an accidental double deletion -- it triggers an undefined behavior in C++ so the program can behave in an arbitrarily weird way. 

Communicator::cleanup gives us a better control. For example, in the new commit, I added a is_available check to at least warn about double cleanup. 

> print a warning otherwise

Good point. Even better, the new commit did `~Communicator() = delete`.